### PR TITLE
Ignore near_plane setting on non-Android platforms

### PR DESCRIPTION
**Pull request by ClobberXD**
`near_plane` seems to be limited to 0.5 to prevent x-ray vision, but even 0.25 simulates x-ray vision across almost 25% of the screen. So this PR:

- Completely ignores this setting on non-Android devices.
- Changes the upper-bound of this setting from 0.5 to 0.25 on Android devices.
- Changes the default value of this setting from 0.1 to 0.05, because the default value would otherwise be the same as the upper-bound.

### To test

- Fire up one instance running the PR branch, and one instance running `master` (for comparison).
- Set `near_plane` to 0.25 in both instances.
- Observe that the x-ray effect is non-existent in the instance running the PR (as the setting is ignored).

Screenshot from 2019-08-04 14-27-55
(left: PR, right: `master`)

****

Needs testing on Android devices to ensure that 0.05 doesn't result in increased visual artifacts.

### CODE REVIEW

File: src/client/camera.cpp [hunk 1]
```diff
@@ -549,8 +549,13 @@ void Camera::updateViewingRange()
 	f32 viewing_range = g_settings->getFloat("viewing_range");
 	f32 near_plane = g_settings->getFloat("near_plane");
 
+#ifdef __ANDROID__
```
Comment (red-001) [inline src/client/camera.cpp]: might be a good idea to leave a comment here on why it's different on android (and presumingly other mobile devices?)

**Comment by Calinou** [comment]
Is there even a reason for this being a setting in minetest.conf? It could be increased automatically on Android and on platforms where the Z-buffer is only 16 bits (i.e. Intel graphics drivers when you don't explicitly request a 24-bit Z-buffer).

**Comment by paramat** [comment]
No need to remove these changes but, there's no need to alter conf.example, as it is regularly auto-generated from settingtypes.txt, so is wasted work.
Just so you know forfuture PRs.

File: src/client/camera.cpp (hunk 1 above)
Comment (ClobberXD) [reply src/client/camera.cpp]: Good idea. Will do.

**Comment by ClobberXD** [comment]
@Calinou I initially did contemplate removing this setting altogether, but dropped the idea because I _thought_ it'd be met with a lot of opposition. :)

Let me know if I should close this PR and _try_ to work on that instead.

@paramat Yes, I'm aware of that, but I did it manually nonetheless, because it's a trivial change, and the file would also be up to date right away.

**Comment by SmallJoker** [comment]
> Is there even a reason for this being a setting in minetest.conf?

Good point. Higher values are only required on less performant hardware, such as the RPI or Android. I would like to see this setting to be changed to something like `increase_near_pane = false` which defaults to `true` on Android. The near pane value would then be 0.1, 0.25 respectively.

**Comment by paramat** [comment]
> Is there even a reason for this being a setting in minetest.conf?

I agree this setting is probably not needed, and i agree with SmallJoker's suggestion (and will +1 it).

ClobberXD, unless you auto-generated conf.example for this PR (doesn't look so because other changes in it would appear) it is highly likely to already be out of date (because most PRs don't bother to update it because it is occasionally auto-generated), so changing it won't make it up to date, so it's wasted work.
It's only worth updating conf.example by auto-generating it, although i'm not requesting that.

**Comment by ClobberXD** [comment]
> I would like to see this setting to be changed to something like `increase_near_plane = false` which defaults to true on Android. The near plane value would then be 0.1, 0.25 respectively.

But that means that anyone could just set it to true to get x-ray vision. So basically nothing would change (except for maybe the lowered upper-bound from 0.5 to 0.25). MT should either auto-detect and auto-configure `near_plane` based on the platform, or use different upper-bounds for different platforms. The latter is probably the easier of the two (not to mention, the PR's already ready :P), and could be merged to make it harder to cheat until the former is implemented.

> it is highly likely to already be out of date

Good point. At least the parts of the file relevant to this PR would be up-to-date, but nevertheless I'll stop updating conf.example in my PRs henceforth. :)

**Comment by benrob0329** [comment]
I think he meant as a compile-time variable.

**Comment by paramat** [comment]
ClobberXD, oh yeah, good point about the setting. I think i misunderstood what SmallJoker meant.
> MT should either auto-detect and auto-configure near_plane based on the platform, or use different upper-bounds for different platforms.

I agree.

**Comment by ClobberXD** [comment]
> I think he meant as a compile-time variable.

I hope that's what he meant. :)

Then we could just detect if the platform is Android, and directly set `near_plane` to 0.25, and remove the setting altogether.

**Comment by ClobberXD** [comment]
Update:

`near_plane` setting is completely ignored on non-Android platforms, and the camera's near-plane is set to a hard-coded value of 0.05. On Android, the behaviour is unchanged (from what the PR originally did).

Btw, shall I move `m_cameranode->setNearValue` to the constructor? I don't see why this is (re)set every time the viewing range is updated.

File: src/client/camera.cpp (hunk 1 above)
Comment (ClobberXD) [reply src/client/camera.cpp]: Added a one-line comment

**Comment by paramat** [comment]
Related comment https://github.com/minetest/minetest/issues/8065#issuecomment-451771466

**Comment by ClobberXD** [comment]
Regarding Calinou's suggestions (both in 8065 and in https://github.com/minetest/minetest/pull/8749#issuecomment-518020711), we can implement automatic near plane configuration later, once this PR has been merged. Note that this PR is sorta high-ish priority, as this setting is used openly to cheat on servers (esp. in survival and PvP modes).

**Comment by ClobberXD** [comment]
Rebased. What's the status of this PR? Also, please note:

> Note that this PR is sorta high-ish priority, as this setting is used openly to cheat on servers (esp. in survival and PvP modes).

**Comment by paramat** [comment]
Added to 5.2.0 milestone.

**Comment by sfan5** [comment]
Since this PR is still in the backlog, here's my feedback:
The problem with the near plane generally applies to weaker GPUs, so it would make sense to replace the `#ifdef __ANDROID__` with `#if ENABLE_GLES`. Other than that the PR looks good to me.

**Comment by paramat** [comment]
Although this is likely to be improved later, shall we merge this for 5.2.0? It is fairly high priority.

**Comment by paramat** [comment]
Testing on desktop PC:
`near_plane = 0.5` is very cheaty, this clientside cheat should not be allowed:

screenshot_20200317_215319

With `near_plane = 0.25` and default FOV, no view inside nodes is possible, so seems a reasonable new upper limit for Android.
`near_plane = 0.05` causes no visual problems for me (Intel integrated GPU).

**Comment by paramat** [comment]
With `near_plane = 0.1`, the current default, and an impractically extreme FOV of 140, viewing inside nodes is not possible. FOV 160 (the maximum FOV) allows an inside view but the world is very distorted so it is not of much use.
So we could safely leave the default at 0.1 for now at least, as there is a concern that 0.05 might cause problems on Android.

So my suggestion is: This PR is fine, but leave the default unchanged at 0.1. Then merge it for 5.2.0.

**Comment by ClobberXD** [comment]
Roger that, will attend to the request in a week's time.

**Comment by paramat** [comment]
A week might be a little late for release, we could merge this and alter it on merge.
I might test the PR, i cannot test on Android but that is not important as nothing really changes, Android is currently unavailable in the Play store anyway.

**Comment by paramat** [comment]
If wanted, i can open a PR that contains ClobberXD's commit, i have created a PR branch https://github.com/paramat/minetest/tree/nearplane

**Comment by sfan5** [comment]
Pushed your commit and the `ENABLE_GLES` change I suggested to the branch of this PR.

**Comment by ClobberXD** [comment]
Oh, are we nearing the release/freeze for 5.2.0? Also, thanks for reverting the min value back to `0.1`.

**Comment by paramat** [comment]
sfan5 thanks.
My branch (before without your commit) works fine, but i will try to test this PR.

**Comment by paramat** [comment]
Tested PR on desktop.
:+1: 
Please can we merge this for 5.2.0?

**Pull request merged**